### PR TITLE
feat: Force Sample Rate for Tracing [DTSS-721]

### PR DESCRIPTION
**Pull request by george-e-shaw-iv**
**What this PR does / why we need it**: 
* Allow traces to force their sample rates.

**JIRA ID**: [DTSS-721]

**Notes for your reviewer**:

[DTSS-721]: https://outreach-io.atlassian.net/browse/DTSS-721

### CODE REVIEW

**Review by bhogatemadhura — APPROVED** [review]

File: pkg/trace/trace.go [hunk 1]
```diff
@@ -198,6 +198,12 @@ func ForceTracing(ctx context.Context) context.Context {
 	return forceTracing(ctx)
 }
 
+// ForceSampleRate will force a desired sample rate for the given trace and all children
+// of said trace. The sample rate in practice will be 1/<rate>.
+func ForceSampleRate(ctx context.Context, rate uint) context.Context {
```
Comment (jaredallard) [inline pkg/trace/trace.go:203]: This is slightly confusing, can we provide a `ExampleSampleRate`?

**Review by jaredallard — APPROVED** [review]
Let's give it a shot :tada:

File: pkg/trace/trace.go (hunk 1 above)
Comment (george-e-shaw-iv) [reply pkg/trace/trace.go:203]: How you'd invoke this in my opinion isn't confusing which is the value of `Example*` functions, but I will add more to that comment. I'll add an explicit example that maths out a sample rate

File: pkg/trace/force_sampler.go [hunk 2]
```diff
@@ -38,3 +47,12 @@ func forceTracing(ctx context.Context) context.Context {
 	}
 	return ctx
 }
+
+// sampleAt sets the sample rate for a given trace explicitly. The rate given will be
+// parsed as 1/rate to actually determine the sample rate.
```
Comment (vk-outreach) [inline pkg/trace/force_sampler.go:52]: nit: the wording here (probably copied from my own from elsewhere) can be improved a little.

```suggestion
// sampleAt sets the sample rate for a given trace explicitly.  
// The rate specifies how many samples were looked at for each 
// accepted sample.  That is, the actual sampling  = 1/rate.
```

File: pkg/trace/force_sampler_test.go [hunk 3]
```diff
@@ -66,3 +66,48 @@ func (suite) TestForceSamplerWhenNotForced(t *testing.T) {
 		}
 	})
 }
+
+func testSampleAtSampler(ctx context.Context, sampleRate uint, handler func(context.Context, *tracetest.TraceLog)) {
+	defer app.SetName(app.Info().Name)
+	app.SetName("gobox")
+
+	trlog := tracetest.NewTraceLogWithOptions(tracetest.Options{
+		SamplePercent: 1,
```
Comment (vk-outreach) [inline pkg/trace/force_sampler_test.go:75]:
```suggestion
		SamplePercent: 0,
```

**Review by vk-outreach — APPROVED** [review]
LGTM

I think this should work

File: pkg/trace/force_sampler.go [hunk 4]
```diff
@@ -38,3 +47,12 @@ func forceTracing(ctx context.Context) context.Context {
...
+func sampleAt(ctx context.Context, rate uint) context.Context {
+	if t := trace.GetTraceFromContext(ctx); t != nil {
+		t.AddField(fieldSampleTrace, rate)
+	}
+	return ctx
+}
```
Comment (vk-outreach) [inline pkg/trace/force_sampler.go:58]: FWIW, I'd not have done it this way but this is ok.

I'd have done it as a log marshaler:

```golang

type TraceSampleRate  uint
func (t TraceSampleRate) MarshalLog(addField func(k string, v interface{})) {
    addField(fieldSampleTrace, rate)
}
```

I believe if we did this way, you can just do:

```golang
    ctx = trace.StartCall(ctx, "foo", log.TraceSampleRate(1000))
```

But we don't expect to use this a whole lot, so it doesn't matter a whole lot.

**Pull request merged**